File: src/adapters/output/messaging/InMemoryEventPublisher.ts
// ============================================================
// Driven Adapter - 記憶體版事件發布者
// ============================================================
// EventPublisher Output Port 的具體實作。
// 在真實場景中，這會被替換為 Kafka、RabbitMQ 等訊息佇列。
//
// 遵循 OCP：新增事件類型時，此類別不需要修改。
// 遵循 LSP：可被任何其他 EventPublisher 實作替換。
//
// 此實作同時負責將寫入模型同步到讀取模型，
// 展示了 CQRS 中事件驅動的讀寫同步機制。
// ============================================================

import { DomainEvent } from '../../../domain/events/DomainEvent';
import { EventPublisher } from '../../../application/ports/output/EventPublisher';
import { InMemoryOrderQueryRepository } from '../persistence/InMemoryOrderQueryRepository';
import { InMemoryOrderCommandRepository } from '../persistence/InMemoryOrderCommandRepository';
import { OrderReadModel } from '../../../application/ports/output/OrderRepository';

export class InMemoryEventPublisher implements EventPublisher {
  private publishedEvents: DomainEvent[] = [];

  constructor(
    private readonly commandRepo: InMemoryOrderCommandRepository,
    private readonly queryRepo: InMemoryOrderQueryRepository,
  ) {}

  async publish(event: DomainEvent): Promise<void> {
    this.publishedEvents.push(event);
    await this.syncReadModel(event.aggregateId);
  }

  async publishAll(events: DomainEvent[]): Promise<void> {
    for (const event of events) {
      await this.publish(event);
    }
  }

  /**
   * 同步讀取模型：將命令端的資料轉換為讀取模型
   * 在真實系統中，這會是一個獨立的事件處理器（Projector）
   */
  private async syncReadModel(orderId: string): Promise<void> {
    const order = await this.commandRepo.findById(orderId);
    if (!order) return;

    const readModel: OrderReadModel = {
      id: order.id,
      customerId: order.customerId,
      items: order.items.map((item) => ({
        productId: item.productId,
        productName: item.productName,
        unitPrice: item.unitPrice,
        quantity: item.quantity,
        subtotal: item.subtotal,
      })),
      totalAmount: order.totalAmount,
      status: order.status,
      createdAt: order.createdAt.toISOString(),
    };

    this.queryRepo.sync(readModel);
  }

  // 測試輔助方法
  getPublishedEvents(): ReadonlyArray<DomainEvent> {
    return [...this.publishedEvents];
  }

  clear(): void {
    this.publishedEvents = [];
  }
}

File: src/main/java/com/example/hexagonal/adapter/output/messaging/InMemoryEventPublisher.java
package com.example.hexagonal.adapter.output.messaging;

// ============================================================
// Driven Adapter - 記憶體版事件發布者
// ============================================================
// 負責發布事件並同步讀取模型（CQRS 讀寫同步）。
// 在真實場景中，可替換為 Kafka / RabbitMQ 實作。
// ============================================================

import com.example.hexagonal.adapter.output.persistence.InMemoryOrderCommandRepository;
import com.example.hexagonal.adapter.output.persistence.InMemoryOrderQueryRepository;
import com.example.hexagonal.application.dto.OrderItemView;
import com.example.hexagonal.application.dto.OrderReadModel;
import com.example.hexagonal.application.port.output.EventPublisher;
import com.example.hexagonal.domain.event.DomainEvent;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class InMemoryEventPublisher implements EventPublisher {

    private final List<DomainEvent> publishedEvents = new ArrayList<>();
    private final InMemoryOrderCommandRepository commandRepo;
    private final InMemoryOrderQueryRepository queryRepo;

    public InMemoryEventPublisher(InMemoryOrderCommandRepository commandRepo,
                                  InMemoryOrderQueryRepository queryRepo) {
        this.commandRepo = commandRepo;
        this.queryRepo = queryRepo;
    }

    @Override
    public void publish(DomainEvent event) {
        publishedEvents.add(event);
        syncReadModel(event.aggregateId());
    }

    @Override
    public void publishAll(List<DomainEvent> events) {
        events.forEach(this::publish);
    }

    /**
     * 同步讀取模型：Command 端 → Read Model。
     * 在真實系統中這會是獨立的 Projector / Event Handler。
     */
    private void syncReadModel(String orderId) {
        commandRepo.findById(orderId).ifPresent(order -> {
            var items = order.getItems().stream()
                    .map(i -> new OrderItemView(
                            i.productId(), i.productName(),
                            i.unitPrice(), i.quantity(), i.subtotal()))
                    .toList();

            var readModel = new OrderReadModel(
                    order.getId(),
                    order.getCustomerId(),
                    items,
                    order.getTotalAmount(),
                    order.getStatus().name(),
                    order.getCreatedAt().toString());

            queryRepo.sync(readModel);
        });
    }

    public List<DomainEvent> getPublishedEvents() {
        return Collections.unmodifiableList(publishedEvents);
    }

    public void clear() { publishedEvents.clear(); }
}
